lib/Pingmachine/Probe/PPing.pm

```diff
@@ -0,0 +1,247 @@
+            my $exit = $cbv->recv;
+            $exit = $exit >> 8;
+            if($exit and $exit != 1 and $exit != 2) {
+                # exit 1 means that some hosts aren't reachable
+                # exit 2 means "any IP addresses were not found"
+                $log->warning("pping seems to have failed (exit: $exit, stderr: ".$job{output}.")");
+                return;
+            }
+
+	    $log->debug("finished: @$cmd (step: $step, pings: $pings, offset: ".$self->time_offset().")") if $log->is_debug();
+
+            $self->_collect_current_job();
+
+	    $log->debug("collected: @$cmd (step: $step, pings: $pings, offset: ".$self->time_offset().")") if $log->is_debug();
+        }
+    );
+}
+
+sub _kill_current_job {
+    my ($self) = @_;
+
+    # Kill pping, if still running
+    my $job = $self->current_job;
+    if($job->{pid}) {
+        if(kill(0, $job->{pid})) {
+            $log->warning("killing unfinished pping process (step: ".$self->step.", pings: ".$self->pings.", offset: ".$self->time_offset().")");
+            kill 9, $job->{pid};
+            $job->{pid} = undef;
+	}
+        elsif($job->{output}) {
+            $log->warning("pping has finished, but we didn't notice... collecting (step: ".$self->step.", pings: ".$self->pings.", offset: ".$self->time_offset.")");
+            $self->_collect_current_job();
+        }
+        else {
+            $log->warning("pping has finished, but we didn't notice... no output found (?)");
+        }
+    }
+}
+
+sub _collect_current_job {
+    my ($self) = @_;
+
+    my $job = $self->current_job;
+    $self->current_job({});
+    my %results;
+
+    # Do nothing, if pping didn't run yet or if job has been already collected
+    return unless $job->{output};
+
+    # Parse pping report
+    my $text = $job->{output};
+    $log->info("Output to parse: text: ".$text."--");
+    while($text !~ /\G\z/gc) {
+        if($text =~ /\G(\S+)[ \t]+:/gc) {
+            my $host = $1;
+            my @data;
+            while($text =~ /\G[ \t]+([-\d\.]+)/gc) {
+                push @data, $1;
+            }
+            # raw ping times
+            my @pings = map {$_ eq '-' ? undef : $_ / 1000} @data;
+            $results{$host}{pings} = \@pings;
+            # sorted rtt times
+            my @rtts = map {sprintf "%.6e", $_ / 1000} sort {$a <=> $b} grep /^\d/, @data;
+            $results{$host}{rtts} = \@rtts;
+            $log->info("adding results: rtts: ".join(" ", @rtts));
+        }
+
+        # discard any other output on the line (ICMP host unreachable errors, etc.)
+	$text =~ /\G.*\n/gc;
+    }
+
+    $log->debug("adding results") if $log->is_debug();
+
+    # Add results (to RRD)
+    if(scalar keys %results) {
+        my $now = int(AnyEvent->now);
+        my $step = $self->step;
+        my $rrd_time = $now - $step - $now%$step;
+        for my $host (keys %results) {
+            my $h2o = $job->{host2order}{$host};
+            if(not defined $h2o) {
```

Comment:
Perl Best Practices: do not use `not`.
And the defined can also be omitted as a reference is expected and all forms of Perl false is false.